import sys
import shutil
import subprocess
import os
import psutil
import ctypes
from PyQt5 import QtWidgets
from ui.mainwindow import Ui_MainWindow   # file UI export t·ª´ Qt Designer


# ‚úÖ L·∫•y ƒë√∫ng th∆∞ m·ª•c user hi·ªán t·∫°i
HOME = os.path.expanduser("~")

# Default Zalo-related folders
FOLDERS = {
    "Zalo":    os.path.join(HOME, "AppData", "Local", "Programs", "Zalo"),
    "ZaloPC":  os.path.join(HOME, "AppData", "Local", "ZaloPC"),
    "ZaloData": os.path.join(HOME, "AppData", "Roaming", "ZaloData"),
    "ZaloUpdate":  os.path.join(HOME, "AppData", "Local", "zalo-updater"),   # ‚úÖ m·ªõi th√™m
}


def get_folder_size(path):
    """Return folder size in MB"""
    total_size = 0
    if not os.path.exists(path):
        return 0
    for dirpath, dirnames, filenames in os.walk(path):
        for f in filenames:
            try:
                fp = os.path.join(dirpath, f)
                if os.path.isfile(fp):
                    total_size += os.path.getsize(fp)
            except Exception:
                pass
    return round(total_size / (1024 * 1024), 2)

def is_junction(path):
    """Check if a folder is a junction/symlink in Windows"""
    if not os.path.exists(path):
        return False
    attrs = ctypes.windll.kernel32.GetFileAttributesW(str(path))
    if attrs == -1:
        return False
    # 0x400 = FILE_ATTRIBUTE_REPARSE_POINT
    return bool(attrs & 0x400)

class ZaloMover(QtWidgets.QMainWindow, Ui_MainWindow):
    def __init__(self):
        super().__init__()
        self.setupUi(self)

        # Connect buttons
        self.browseButton.clicked.connect(self.choose_folder)
        self.moveButton.clicked.connect(self.move_selected)
        self.deleteButton.clicked.connect(self.delete_old_backups)

        # Reset progress bar
        self.progressBar.setValue(0)

        # Set app title
        self.setWindowTitle("ZaloMove - Ph√°t tri·ªÉn b·ªüi Shun")

        # ‚úÖ Vi·ªát h√≥a label n√∫t
        self.browseButton.setText("Ch·ªçn th∆∞ m·ª•c...")
        self.moveButton.setText("Di chuy·ªÉn th∆∞ m·ª•c Zalo")

        # ‚úÖ Disable checkbox n·∫øu folder kh√¥ng t·ªìn t·∫°i v√† show size n·∫øu c√≥
        self.check_folders()

    def check_folders(self):
        """Disable checkboxes if folder kh√¥ng t·ªìn t·∫°i ho·∫∑c ƒë√£ l√† symbolic link"""
        for name, path in FOLDERS.items():
            size = get_folder_size(path)
            label = f"{name} ({size} MB)" if size > 0 else f"{name} (Not Found)"

            disabled = (not os.path.exists(path)) or is_junction(path)

            if name == "Zalo":
                self.checkZalo.setText(label)
                self.checkZalo.setEnabled(not disabled)
            elif name == "ZaloPC":
                self.checkZaloPC.setText(label)
                self.checkZaloPC.setEnabled(not disabled)
            elif name == "ZaloData":
                self.checkZaloData.setText(label)
                self.checkZaloData.setEnabled(not disabled)
            elif name == "ZaloUpdate":   # ‚úÖ x·ª≠ l√Ω th√™m checkbox m·ªõi
                self.checkZaloUpdate.setText(label)
                self.checkZaloUpdate.setEnabled(not disabled)

    def choose_folder(self):
        """Open folder chooser dialog"""
        folder = QtWidgets.QFileDialog.getExistingDirectory(self, "Ch·ªçn th∆∞ m·ª•c ƒë√≠ch m·ªõi")
        if folder:
            self.newPath.setText(folder)

    def is_zalo_running(self):
        """Check if Zalo is running"""
        for proc in psutil.process_iter(['name']):
            try:
                if proc.info['name'] and "Zalo" in proc.info['name']:
                    return True
            except (psutil.NoSuchProcess, psutil.AccessDenied):
                continue
        return False

    def kill_zalo(self):
        """Force kill all Zalo processes"""
        killed = 0
        for proc in psutil.process_iter(['name']):
            try:
                if proc.info['name'] and "Zalo" in proc.info['name']:
                    proc.kill()
                    killed += 1
            except (psutil.NoSuchProcess, psutil.AccessDenied):
                continue
        return killed

    def move_selected(self):
        """Move selected folder(s)"""
        user_base = self.newPath.text().strip()

        if not user_base:
            QtWidgets.QMessageBox.warning(self, "C·∫£nh b√°o", "Vui l√≤ng ch·ªçn th∆∞ m·ª•c ƒë√≠ch tr∆∞·ªõc.")
            return

        # üîí Disable n√∫t move ƒë·ªÉ tr√°nh b·∫•m nhi·ªÅu l·∫ßn
        self.moveButton.setEnabled(False)
        self.moveButton.setText("ƒêang x·ª≠ l√Ω...")

        try:
            # N·∫øu Zalo ƒëang ch·∫°y ‚Üí t·ª± kill
            if self.is_zalo_running():
                killed = self.kill_zalo()
                QtWidgets.QMessageBox.information(self, "Th√¥ng b√°o", f"Zalo ƒëang b·ªã ƒë√≥ng ƒë·ªÉ di chuy·ªÉn, nh·∫•n OK ƒë·ªÉ ti·∫øp t·ª•c")

            # ‚úÖ Always create 'ZaloMove' inside the chosen folder
            new_base = os.path.join(user_base, "ZaloMove")
            os.makedirs(new_base, exist_ok=True)

            # Determine which folders are checked
            selected = []
            if self.checkZalo.isChecked() and self.checkZalo.isEnabled():
                selected.append("Zalo")
            if self.checkZaloPC.isChecked() and self.checkZaloPC.isEnabled():
                selected.append("ZaloPC")
            if self.checkZaloData.isChecked() and self.checkZaloData.isEnabled():
                selected.append("ZaloData")
            if self.checkZaloUpdate.isChecked() and self.checkZaloUpdate.isEnabled():   # ‚úÖ m·ªõi th√™m
                selected.append("ZaloUpdate")

            if not selected:
                QtWidgets.QMessageBox.warning(self, "C·∫£nh b√°o", "Vui l√≤ng ch·ªçn √≠t nh·∫•t m·ªôt th∆∞ m·ª•c h·ª£p l·ªá.")
                return

            total = len(selected)
            self.progressBar.setMaximum(total)
            self.progressBar.setValue(0)

            errors = []

            for i, name in enumerate(selected, start=1):
                old_path = FOLDERS[name]
                new_path = os.path.join(new_base, name)

                # N·∫øu folder ƒë√£ t·ªìn t·∫°i trong zalo_move ‚Üí h·ªèi c√≥ overwrite kh√¥ng
                if os.path.exists(new_path):
                    reply = QtWidgets.QMessageBox.question(
                        self,
                        "Th∆∞ m·ª•c ƒë√£ t·ªìn t·∫°i",
                        f"{new_path} ƒë√£ t·ªìn t·∫°i. B·∫°n c√≥ mu·ªën ghi ƒë√® kh√¥ng?",
                        QtWidgets.QMessageBox.Yes | QtWidgets.QMessageBox.No,
                        QtWidgets.QMessageBox.No
                    )
                    if reply == QtWidgets.QMessageBox.No:
                        continue
                    shutil.rmtree(new_path, ignore_errors=True)

                if not os.path.exists(old_path):
                    errors.append(f"{name} kh√¥ng t√¨m th·∫•y t·∫°i {old_path}")
                else:
                    try:
                        # T·∫°o b·∫£n sao l∆∞u tr∆∞·ªõc khi di chuy·ªÉn: <folder>.old
                        backup_path = f"{old_path}.old"
                        if os.path.exists(backup_path):
                            reply_backup = QtWidgets.QMessageBox.question(
                                self,
                                "Backup ƒë√£ t·ªìn t·∫°i",
                                f"ƒê√£ c√≥ b·∫£n sao l∆∞u: {backup_path}. B·∫°n c√≥ mu·ªën ghi ƒë√® kh√¥ng?",
                                QtWidgets.QMessageBox.Yes | QtWidgets.QMessageBox.No,
                                QtWidgets.QMessageBox.No
                            )
                            if reply_backup == QtWidgets.QMessageBox.No:
                                # B·ªè qua th∆∞ m·ª•c n√†y n·∫øu kh√¥ng mu·ªën ghi ƒë√® backup
                                continue
                            shutil.rmtree(backup_path, ignore_errors=True)

                        shutil.copytree(old_path, backup_path)

                        # Move folder
                        shutil.move(old_path, new_path)

                        # Create symlink (junction)
                        subprocess.run(f'mklink /J "{old_path}" "{new_path}"',
                                    shell=True, check=True)

                    except Exception as e:
                        errors.append(f"L·ªói khi x·ª≠ l√Ω {name}: {e}")

                # Update progress bar
                self.progressBar.setValue(i)
                QtWidgets.QApplication.processEvents()

            if errors:
                QtWidgets.QMessageBox.critical(self, "K·∫øt qu·∫£", "\n".join(errors))
            else:
                QtWidgets.QMessageBox.information(self, "Th√†nh c√¥ng", f"ƒê√£ di chuy·ªÉn: {', '.join(selected)}")

        finally:
            # üîì Enable l·∫°i n√∫t move khi xong
            self.moveButton.setEnabled(True)
            self.moveButton.setText("Di chuy·ªÉn th∆∞ m·ª•c Zalo")

    def delete_old_backups(self):
        """X√≥a c√°c th∆∞ m·ª•c backup (*.old) n·∫øu t·ªìn t·∫°i"""
        # T√¨m t·∫•t c·∫£ backup t·ªìn t·∫°i
        backups = []
        for name, path in FOLDERS.items():
            backup_path = f"{path}.old"
            if os.path.exists(backup_path):
                backups.append((name, backup_path))

        if not backups:
            QtWidgets.QMessageBox.information(self, "Th√¥ng b√°o", "Kh√¥ng t√¨m th·∫•y th∆∞ m·ª•c backup (.old) n√†o ƒë·ªÉ x√≥a.")
            return

        # H·ªèi x√°c nh·∫≠n
        backup_list_text = "\n".join([bp for _, bp in backups])
        reply = QtWidgets.QMessageBox.question(
            self,
            "X√°c nh·∫≠n x√≥a backup",
            f"B·∫°n c√≥ ch·∫Øc ch·∫Øn mu·ªën x√≥a c√°c th∆∞ m·ª•c backup sau?\n\n{backup_list_text}",
            QtWidgets.QMessageBox.Yes | QtWidgets.QMessageBox.No,
            QtWidgets.QMessageBox.No
        )
        if reply == QtWidgets.QMessageBox.No:
            return

        errors = []
        deleted = []
        for name, backup_path in backups:
            try:
                shutil.rmtree(backup_path, ignore_errors=False)
                deleted.append(name)
            except Exception as e:
                errors.append(f"L·ªói khi x√≥a backup {name}: {e}")

        if errors:
            QtWidgets.QMessageBox.critical(self, "K·∫øt qu·∫£", "\n".join(errors))
        else:
            QtWidgets.QMessageBox.information(self, "Th√†nh c√¥ng", f"ƒê√£ x√≥a backup: {', '.join(deleted)}")


if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    window = ZaloMover()
    window.show()
    sys.exit(app.exec_())
